client/src/app/hooks/useAssessmentStatus.ts

```diff
@@ -0,0 +1,77 @@
+// hooks/useAssessmentStatus.js
+import { Assessment, Archetype, Application } from "@app/api/models";
+import { useFetchArchetypes } from "@app/queries/archetypes";
+import { useFetchAssessments } from "@app/queries/assessments";
+import { useMemo } from "react";
+
+export const useAssessmentStatus = (application: Application) => {
+  const { assessments } = useFetchAssessments();
+  const { archetypes } = useFetchArchetypes();
+
+  return useMemo(() => {
+    const applicationAssessments = assessments?.filter(
+      (assessment: Assessment) => assessment.application?.id === application.id
+    );
+    const inheritedArchetypes = archetypes?.filter(
+      (archetype: Archetype) =>
+        archetype.applications?.map((app) => app.id).includes(application.id)
+    );
+
+    const assessmentsWithArchetypes = inheritedArchetypes.map(
+      (inheritedArchetype) => ({
+        inheritedArchetype,
+        assessments: assessments.filter(
+          (assessment) => assessment.archetype?.id === inheritedArchetype.id
+        ),
+      })
+    );
+
+    const allArchetypesAssessed =
+      assessmentsWithArchetypes.length > 0 &&
+      assessmentsWithArchetypes.every(({ inheritedArchetype, assessments }) => {
+        const requiredAssessments = assessments.filter(
+          (assessment) => assessment.required
+        );
+        return (
+          inheritedArchetype.assessed &&
+          assessments.length > 0 &&
+          requiredAssessments.length > 0 &&
+          requiredAssessments.every(
+            (assessment) => assessment.status === "complete"
+          )
+        );
+      });
+
+    const assessmentsFromArchetypesCount = assessmentsWithArchetypes.filter(
+      ({ assessments }) => assessments.some((assessment) => assessment.required)
+    ).length;
+
+    const assessedArchetypesCount = assessmentsWithArchetypes.filter(
+      ({ assessments, inheritedArchetype }) =>
+        assessments.some(
+          (assessment) =>
+            assessment.required &&
+            assessment.status === "complete" &&
+            inheritedArchetype.assessed
+        )
+    ).length;
+
+    const hasApplicationAssessmentInProgress = applicationAssessments?.some(
+      (assessment: Assessment) =>
+        assessment.status === "started" ||
+        assessment.status === "empty" ||
+        assessment.status === "complete"
+    );
+    const isDirectlyAssessed =
+      (application.assessed && (application.assessments?.length ?? 0) > 0) ??
+      false;
+
+    return {
+      allArchetypesAssessed,
+      countOfFullyAssessedArchetypes: assessedArchetypesCount,
+      countOfArchetypesWithRequiredAssessments: assessmentsFromArchetypesCount,
+      hasApplicationAssessmentInProgress,
+      isApplicationDirectlyAssessed: isDirectlyAssessed,
+    };
+  }, [assessments, archetypes, application]);
```

Comment:
Right now the memo will get re-calculated with every react query re-fetch of all three entities involved (assuming that `application` reference is stable between re-fetch).
We could avoid re-calculating due to applications by  providing only `application.id`  and `isDirectlyAssessed` as a dependencies.

Reply:
I'm ok with this - we do have 
eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)
turned on so wondering if we are all on board with an exception here.

Reply:
On second thought.. the unit tests broke when I removed the other dependencies from the array.

Reply:
This seems to be one of the times where `useMemo` can be helpful if all the things don't change, and adds minimal overhead if it changes a bunch.  I'd keep it the way it is.

Also, I don't think the refetches will reset the returned data arrays on every fetch.  I seem to remember there is some caching built in and the fetched data != the returned data for this kind of reason.